students/JerryH/Lesson09/mailroom.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
+
+class Donor:
+
+    def __init__(self, first, last, donations = None):
+        self.first = first
+        self.last = last
+        self.donations = donations
+
+    @property
+    def full_name(self):
+        return "{} {}".format(self.first, self.last)
+
+    @property
+    def donor_history(self):
+        return sum(self.donations)
+
+    def add_donation(self, amount):
+        return self.donations.append(amount)
+
+
+class DonorBook:
+
+    def __init__(self, donors = []):
+        if donors is None:
+            self.donors = []
+        else:
+            self.donors = donors
+
+    def add_donor(self, donor):
+        # if donor not in self.donors:
+        self.donors.append(donor)
+
+    def get_all_donor_names(self):
+        donor_name_list = []
+        for donor in self.donors:
+            #print("-->" + donor.full_name)
+            donor_name_list.append(donor.full_name)
+        return donor_name_list
+
+    def list_all_donor_names_sorted(self):
+        all_names = ""
+        for each_donor_name in sorted(self.get_all_donor_names()):
+            all_names += "{}\n".format(each_donor_name)
+        return(all_names)
+
+    def send_thank_you(self):
+        donor_name = None
+        while not donor_name:
+            donor_name = donor_name_prompt()
+            if donor_name.lower() == "list":
+                print(self.list_all_donor_names_sorted())
+                donor_name = None
+
+        donation = None
+        while not donation:
+            try:
+                donation = donation_prompt()
+            except ValueError:
+                print("Not a valid number! Please enter a valid number:\n")
+
+        # If the donnor doesn't exist in the donor list - add his info
+        if donor_name not in self.get_all_donor_names():
+            try:
+                first, last = donor_name.split(" ")
+                self.add_donor(Donor(first, last, [donation]))
+            except ValueError:
+                print("Please enter both of your \"First Name\" and \"Last Name\"")
+        else:
+            for donor in self.donors:
+                if donor.full_name == donor_name:
+                    donor.add_donation(donation)
+
+        print("Thank You Email:  Thanks for the donation!\n\n")
```

Comment:
it is best to extract some of this code outside of the class, specifically one dealing with user interaction